bin/varnishd/cache/cache_range.c

```diff
@@ -180,13 +181,66 @@ vrg_dorange(struct req *req, const char *r, void **priv)
 	return (NULL);
 }
 
+/*
+ * return 1 if range should be observed, based on if-range value
+ * if-range can either be a date or an ETag [RFC7233 3.1 p8]
+ */
+static int
+vrg_ifrange(struct req *req)
+{
+	const char *p, *e;
+	vtim_real ims, lm, d;
+
+	if (!http_GetHdr(req->http, H_If_Range, &p))
+		return (1);
+
+	/* strong validation needed */
+	if (p[0] == 'W' && p[1] == '/')
+		return (0);
+
+	/* ETag */
+	if (p[0] == '"') {
+		if (!http_GetHdr(req->resp, H_ETag, &e))
+			return (0);
+		if ((e[0] == 'W' && e[1] == '/'))
+			return (0);
+		return (strcmp(p, e) == 0);
+	}
+
+	/* assume date, strong check [RFC7232 2.2.2 p7] */
+	if (!(ims = VTIM_parse(p)))
+		return (0);
+
+	/* the response needs a Date */
+	if (!http_GetHdr(req->resp, H_Date, &p) || !(d = VTIM_parse(p)))
+		return (0);
+
+	/* grab the Last Modified value */
+	if (http_GetHdr(req->resp, H_Last_Modified, &p)) {
+		lm = VTIM_parse(p);
+		if (!lm)
+			return (0);
+	} else
+		AZ(ObjGetDouble(req->wrk, req->objcore, OA_LASTMODIFIED, &lm));
```

Comment:
do we really want this? What if the LM header was removed in VCL on purpose to force the If-Range check to only ever work on Etags?

Reply:
I was (and still am) on the fence about this, I mimicked what `RFC2616_Do_Cond()` does (https://github.com/varnishcache/varnish-cache/blob/master/bin/varnishd/cache/cache_rfc2616.c#L296).

Reply:
Interesting, continued in #3118